File: remote/src/main/java/com/kelsos/mbrc/ui/fragments/MiniControlFragment.java
package com.kelsos.mbrc.ui.fragments;

import android.graphics.Bitmap;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;
import butterknife.Bind;
import butterknife.ButterKnife;
import butterknife.OnClick;
import com.google.inject.Inject;
import com.kelsos.mbrc.R;
import com.kelsos.mbrc.annotations.PlayerState;
import com.kelsos.mbrc.presenters.MiniControlPresenter;
import com.kelsos.mbrc.ui.views.MiniControlView;
import com.kelsos.mbrc.utilities.FontUtils;
import roboguice.fragment.RoboFragment;

public class MiniControlFragment extends RoboFragment implements MiniControlView {

  @Bind(R.id.mc_track_cover) ImageView trackCover;
  @Bind(R.id.mc_track_artist) TextView trackArtist;
  @Bind(R.id.mc_track_title) TextView trackTitle;
  @Bind(R.id.mc_play_pause) ImageButton playPause;

  @Inject private MiniControlPresenter presenter;

  @NonNull public static Fragment newInstance() {
    return new MiniControlFragment();
  }

  @OnClick(R.id.mc_next_track) public void onNextClicked() {
    presenter.onNextPressed();
  }

  @OnClick(R.id.mc_play_pause) public void onPlayPauseClicked() {
    presenter.onPlayPause();
  }

  @OnClick(R.id.mc_prev_track) public void onPreviousClicked() {
    presenter.onPreviousPressed();
  }

  @Override public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
    final View view = inflater.inflate(R.layout.ui_fragment_mini_control, container, false);
    ButterKnife.bind(this, view);
    presenter.bind(this);
    trackTitle.setTypeface(FontUtils.getRobotoMedium(getActivity()));
    trackArtist.setTypeface(FontUtils.getRobotoRegular(getActivity()));
    presenter.load();
    return view;
  }

  @Override public void onPause() {
    super.onPause();
    presenter.onPause();
  }

  @Override public void onResume() {
    super.onResume();
    presenter.onResume();
  }

  public void updateCover(@Nullable Bitmap cover) {
    if (cover != null) {
      trackCover.setImageBitmap(cover);
    } else {
      trackCover.setImageResource(R.drawable.ic_image_no_cover);
    }
  }

  public void updateTrack(String artist, String title) {
    trackArtist.setText(artist);
    trackTitle.setText(title);
  }

  public void updatePlayerState(@PlayerState.State String playState) {
    switch (playState) {
      case PlayerState.PLAYING:
        playPause.setImageResource(R.drawable.ic_pause_black_36dp);
        break;
      default:
        playPause.setImageResource(R.drawable.ic_play_arrow_black_36dp);
        break;
    }
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/ui/mini_control/MiniControlFragment.kt
package com.kelsos.mbrc.ui.mini_control

import android.content.Intent
import android.graphics.Bitmap
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.app.TaskStackBuilder
import androidx.fragment.app.Fragment
import com.kelsos.mbrc.R
import com.kelsos.mbrc.content.active_status.PlayerState
import com.kelsos.mbrc.content.active_status.PlayerState.State
import com.kelsos.mbrc.content.library.tracks.TrackInfo
import com.kelsos.mbrc.databinding.UiFragmentMiniControlBinding
import com.kelsos.mbrc.extensions.getDimens
import com.kelsos.mbrc.ui.navigation.main.MainActivity
import com.squareup.picasso.Picasso
import toothpick.Toothpick
import java.io.File
import javax.inject.Inject

class MiniControlFragment : Fragment(), MiniControlView {

  private var _binding: UiFragmentMiniControlBinding? = null
  private val binding get() = _binding!!

  @Inject
  lateinit var presenter: MiniControlPresenter

  private fun onControlClick() {
    val builder = TaskStackBuilder.create(requireContext())
    builder.addNextIntentWithParentStack(Intent(context, MainActivity::class.java))
    builder.startActivities()
  }

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
    binding.miniControl.setOnClickListener { onControlClick() }
    binding.mcNextTrack.setOnClickListener { presenter.next() }
    binding.mcPlayPause.setOnClickListener { presenter.playPause() }
    binding.mcPrevTrack.setOnClickListener { presenter.previous() }
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    val scope = Toothpick.openScopes(requireActivity().application, this)
    scope.installModules(MiniControlModule())
    super.onCreate(savedInstanceState)
    Toothpick.inject(this, scope)
  }

  override fun onCreateView(
    inflater: LayoutInflater,
    container: ViewGroup?,
    savedInstanceState: Bundle?
  ): View {
    _binding = UiFragmentMiniControlBinding.inflate(inflater)
    return binding.root
  }

  override fun onStart() {
    super.onStart()
    presenter.attach(this)
    presenter.load()
  }

  override fun onStop() {
    super.onStop()
    presenter.detach()
  }

  override fun updateCover(path: String) {
    val file = File(path)

    if (file.exists()) {
      val dimens = requireContext().getDimens()
      Picasso.get()
        .load(file)
        .noFade()
        .config(Bitmap.Config.RGB_565)
        .resize(dimens, dimens)
        .centerCrop()
        .into(binding.mcTrackCover)
    } else {
      binding.mcTrackCover.setImageResource(R.drawable.ic_image_no_cover)
    }
  }

  override fun updateTrackInfo(trackInfo: TrackInfo) {
    binding.mcTrackArtist.text = trackInfo.artist
    binding.mcTrackTitle.text = trackInfo.title
  }

  override fun updateState(@State state: String) {
    when (state) {
      PlayerState.PLAYING -> binding.mcPlayPause.setImageResource(R.drawable.ic_action_pause)
      else -> binding.mcPlayPause.setImageResource(R.drawable.ic_action_play)
    }
  }

  override fun onDestroy() {
    Toothpick.closeScope(this)
    super.onDestroy()
  }

  @javax.inject.Scope
  @Retention(AnnotationRetention.RUNTIME)
  annotation class Presenter
}
